primitives/src/fee_handler.rs

```
@@ -0,0 +1,119 @@
+}
+
+impl<AccountId, CurrencyHandler, Affiliate, Aid, Tournament, Tid, Treasury> FeeHandler
+	for GameFeeHandler<AccountId, CurrencyHandler, Affiliate, Tournament, Treasury>
+where
+	AccountId: Parameter,
+	CurrencyHandler: Currency<AccountId>,
+	Affiliate: FeeProvider<
+		AccountId = AccountId,
+		FeeIdentifier = Aid,
+		FeeCurrency = CurrencyHandler::Balance,
+		FeeOutput = Vec<(CurrencyHandler::Balance, AccountId)>,
+	>,
+	Aid: Parameter,
+	Tournament: FeeProvider<
+		AccountId = AccountId,
+		FeeIdentifier = Tid,
+		FeeCurrency = CurrencyHandler::Balance,
+		FeeOutput = (CurrencyHandler::Balance, AccountId),
+	>,
+	Tid: Parameter,
+	Treasury: TreasuryManager<AccountId = AccountId, Currency = CurrencyHandler::Balance>,
+{
+	type AccountId = AccountId;
+	type FeeCurrency = CurrencyHandler::Balance;
+	type AffiliateFeeIdentifier = Aid;
+	type TournamentFeeIdentifier = Tid;
+	type TreasuryKey = Treasury::TreasuryKey;
+
+	fn try_propagate_chain_fee(
+		base_fee: Self::FeeCurrency,
+		account: &Self::AccountId,
+		identifier: &Self::AffiliateFeeIdentifier,
+	) -> Result<Self::FeeCurrency, DispatchError> {
+		let mut final_fee = base_fee;
+
+		for (transfer_fee, chain_account) in Affiliate::get_fee_from(base_fee, account, identifier)
+		{
+			if transfer_fee > 0_u32.into() {
+				CurrencyHandler::transfer(account, &chain_account, transfer_fee, AllowDeath)?;
+				final_fee = final_fee.saturating_sub(transfer_fee);
+			}
+		}
+
+		Ok(final_fee)
+	}
+
+	fn try_propagate_tournament_fee(
+		base_fee: Self::FeeCurrency,
+		account: &Self::AccountId,
+		identifier: &Self::TournamentFeeIdentifier,
+	) -> Result<Self::FeeCurrency, DispatchError> {
+		let (tournament_fee, tournament_account) =
+			Tournament::get_fee_from(base_fee, account, identifier);
+
+		if tournament_fee > 0_u32.into() {
+			CurrencyHandler::transfer(account, &tournament_account, tournament_fee, AllowDeath)?;
+			Ok(base_fee.saturating_sub(tournament_fee))
```

Just to be sure, I think I would prefer to return an error here too.